### ISSUE TITLE
Incorrect error in factory function on a dataclass

### ISSUE THREAD

**Issue by wkschwartz** [python/mypy#5263]
In Python `3.7.0rc1` and mypy `0.620+dev-5dc7454b1acb47dcbb3361c849e6113efcdbaaf2`, if I save the following code to `dc_test.py`

```python
from typing import Type, TypeVar
from dataclasses import dataclass

T = TypeVar('T', bound='A')

@dataclass
class A:
	@classmethod
	def make(cls: Type[T]) -> T:
		return cls()
```

I get the following mypy output

```
$ mypy dc_test.py 
dc_test.py:10: error: Incompatible return value type (got "A", expected "T")
```

If I comment out the `@dataclass` line, I get no error. I would expect to get no error with or without the `@dataclass` line.

**Comment by gvanrossum** [comment]
Confirmed. All I can think of is that something's wrong with how class decorators are implemented in mypy.

**Comment by ilevkivskyi** [comment]
The type first argument of class methods is patched by plugins for `attrs` and `dataclases` (because `__init__` is auto-generated). It looks like this patching is not done carefully (it needs to interact well with `bind_self`).

**Comment by wkschwartz** [comment]
# Workaround until this issue is fixed

Casting the result to type `T` silences mypy's false positive error.

```diff
-from typing import Type, TypeVar
+from typing import Type, TypeVar, cast
 from dataclasses import dataclass

 T = TypeVar('T', bound='A')

 @dataclass
 class A:
 	@classmethod
 	def make(cls: Type[T]) -> T:
-		return cls()
+		return cast(T, cls())
```
